File: fsharp-2017/src/Day06.fs
module Day06

open Util

let zeroAt (i:int) (a:int[]) =
  [| for j in 0 .. a.Length-1 -> if j=i then 0 else a.[j] |]

/// Allocate n to all elms in a
let allocAll (n:int) (a:int[]) = [| for x in a -> x + n |]

/// Allocate 1 to n elms in a, starting from elm after idx
let allocN (idx:int) (n:int) (a:int[]) = 
  if n = 0 then a else
  let zeroOrOne i = // 0 or 1, based on index i in the array
    let toEnd = a.Length - idx - 1
    if (i > idx && i - idx <= n) || (i < idx && i + toEnd < n) then 1 else 0
  [| for i in 0 .. a.Length-1 -> a.[i] + zeroOrOne i |]

let hash a = sprintf "%A" <| a

let alloc a:int[] =
  let max = Array.max a
  let i = a |> Array.findIndex (fun x -> x = max)
  a
  |> zeroAt i
  |> allocAll (max / a.Length |> int)
  |> allocN i (max % a.Length)

/// Number of allocations needed for array a
let rec cntAllocs (a:int[]) (hist:Set<string>) = 
  if hist.Contains(hash a) then hist.Count else cntAllocs (alloc a) (hist.Add(hash a))

/// Number of allocations between first time array a is repeated
let rec cntPrvSeen (a:int[]) (hist:Map<string, int>) = 
  let iSeen = hist.TryFind (hash a)
  match iSeen with
  | Some i' -> hist.Count - i'
  | None -> cntPrvSeen (alloc a) (hist.Add(hash a, hist.Count))

let run (instr:string) =
  printfn "Day 6-1 answer: %A" <| cntAllocs (instr |> splitInt '\t') Set.empty
  printfn "Day 6-2 answer: %A" <| cntPrvSeen (instr |> splitInt '\t') Map.empty

File: js-pipe-fkit_2017/src/day06.js
const { maximum } = require('fkit')
const { objLen, strToInts } = require('./helpers')

const zeroAt = i => a /*int,[int]*/ => a.map((x, j) => j==i? 0: x)
const mapAdd = n => a /*int,[int]*/ => a.map(x => x + n)

/// Add 1 to n elms in a, starting from elm __after__ i 
const addOne = (i, n) => a /*int,int,[int]*/ =>
  n == 0
  ? a
  : a.map((x, j) => {
    const tilEnd = a.length - i - 1
    const _01 = (j > i && j-i <= n) || (j < i && j+tilEnd < n)? 1 : 0
    return x + _01
  })

const hash = a => `${a}`
const setHash = (hist, val, a) => {
  hist[hash(a)] = val
  return hist
}

const alloc = a /*[int]*/ => {
  const max = maximum(a)
  const i = a.findIndex(x => x == max)
  return a
  |> zeroAt(i)
  |> mapAdd(~~(max / a.length))
  |> addOne(i, max % a.length)
}

/// returns Number of allocations needed for array a
const countAllocs = (a, hist) /*[int],{}*/ =>
  hist[hash(a)]? objLen(hist): countAllocs(alloc(a), setHash(hist, true, a))

/// return Number of allocations between first time array a is repeated
const countPrvSeen = (a, hist) /*[int],{}*/ => {
  const i = hist[hash(a)]
  return i != null
    ? objLen(hist)-i
    : countPrvSeen(alloc(a), setHash(hist, objLen(hist), a))
}

const run = (inStr, log) => {
  const ints = strToInts(inStr, '\t')
  log.p1( countAllocs(ints, {}) )  // 3156
  log.p2( countPrvSeen(ints, {}) ) // 1610
}
module.exports = {
  run,
  countAllocs,
  countPrvSeen,
}